compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -3500,25 +3500,29 @@ private void validateAnnotationAttachmentExpr(BLangAnnotationAttachment annAttac
             return;
         }
 
-
         // At this point the type is a subtype of  map<anydata>|record{ anydata...; } or
-        // map<anydata>[]|record{ anydata...; }[], thus an expression is required.
-        if (annAttachmentNode.expr == null) {
+        // map<anydata>[]|record{ anydata...; }[]. For map<anydata>[]|record{ anydata...; }[] an expression is
+        // required while map<anydata>|record{ anydata...; } can be empty.
+        if (annAttachmentNode.expr == null && annotationSymbol.attachedType != null &&
+                !(annotationSymbol.attachedType.type.tag == TypeTags.MAP ||
+                        annotationSymbol.attachedType.type.tag == TypeTags.RECORD)) {
             this.dlog.error(annAttachmentNode.pos, DiagnosticErrorCode.ANNOTATION_ATTACHMENT_REQUIRES_A_VALUE,
-                            annotationSymbol);
+                    annotationSymbol);
```

Comment:
Is the assumption regarding arrays here correct? The spec says 

> For an annotation tag declared with type T, **if T is M[] for some type mapping type M, then the type of the annotation value is M**, multiple annotations are allowed, and the effective value is an array of the values for each annotation; otherwise the type of the annotation value is T, multiple annotations are not allowed, the effective value is the value of the single annotation.

and

> If the type of the annotation value is true, then a mapping-constructor-expr is not allowed, and the annotation value is true. **Otherwise the type of the annotation value must be a mapping type M; if a mapping-constructor-expr is not specified, then it defaults to { }**; the mapping-constructor-expr must have static type M and the annotation value is the result of evaluating the mapping-constructor-expr with M as the contextually expected type.

So even in the array scenario, the type of the annotation value can be a mapping for which there is no mapping constructor expression specified, right?

Reply:
Basically, shouldn't the following work?

```ballerina
annotation map<int>[] x on function;

@x
@x
function func() {
} 
```